File: packages/domain/src/public/categories.ts
/**
 * The closed list (ADR-0003). Events at places; nothing about a person. The
 * list cannot be extended at runtime and the test asserts there is no
 * category whose name contains a word for a person.
 */
export const CATEGORIES = [
  'robbery',
  'burglary',
  'road_blocked',
  'accident',
  'fire',
  'flooding',
  'gunfire_heard',
  'unrest_or_protest',
  'building_collapse',
  'power_line_down',
  'missing_person_appeal',
] as const;

export type Category = (typeof CATEGORIES)[number];

/** Hours before a report of each category leaves the feed (FR-4.5). */
export const EXPIRY_HOURS: Readonly<Record<Category, number>> = {
  robbery: 24,
  burglary: 24,
  road_blocked: 6,
  accident: 6,
  fire: 12,
  flooding: 48,
  gunfire_heard: 6,
  unrest_or_protest: 12,
  building_collapse: 48,
  power_line_down: 24,
  missing_person_appeal: 72,
};

/**
 * The one category that involves a person, available only to a verified
 * organisation or verified next of kin, and never distributed without a
 * human's review.
 */
export const HUMAN_REVIEW_ALWAYS: ReadonlySet<Category> = new Set(['missing_person_appeal']);

export function isCategory(s: string): s is Category {
  return (CATEGORIES as ReadonlyArray<string>).includes(s);
}

export function expired(category: Category, reportedMinutes: number, nowMinutes: number): boolean {
  return nowMinutes - reportedMinutes >= EXPIRY_HOURS[category] * 60;
}

/** Per account: three per day and one per half hour (FR-4.6). */
export function mayReportAgain(previousReportMinutes: ReadonlyArray<number>, nowMinutes: number): boolean {
  const day = previousReportMinutes.filter((m) => nowMinutes - m < 24 * 60).length;
  const halfHour = previousReportMinutes.some((m) => nowMinutes - m < 30);
  return day < 3 && !halfHour;
}

File: server/src/Sentinel.Domain/Categories.cs
namespace Sentinel.Domain;

/// <summary>
/// The closed list (ADR-0003): events at places, nothing about a person, and
/// the hours before each leaves the feed. Held to the TypeScript by
/// <c>fixtures/categories.json</c>.
/// </summary>
public static class Categories
{
    public static readonly IReadOnlyDictionary<string, int> ExpiryHours = new Dictionary<string, int>
    {
        ["robbery"] = 24,
        ["burglary"] = 24,
        ["road_blocked"] = 6,
        ["accident"] = 6,
        ["fire"] = 12,
        ["flooding"] = 48,
        ["gunfire_heard"] = 6,
        ["unrest_or_protest"] = 12,
        ["building_collapse"] = 48,
        ["power_line_down"] = 24,
        ["missing_person_appeal"] = 72,
    };

    /// <summary>The one category that involves a person: a verified organisation only, and never without a human's review.</summary>
    public static readonly IReadOnlySet<string> HumanReviewAlways = new HashSet<string> { "missing_person_appeal" };

    public static bool IsCategory(string s) => ExpiryHours.ContainsKey(s);

    public static bool Expired(string category, long reportedMinutes, long nowMinutes) =>
        nowMinutes - reportedMinutes >= ExpiryHours[category] * 60L;

    /// <summary>Per account: three per day and one per half hour (FR-4.6).</summary>
    public static bool MayReportAgain(IEnumerable<long> previousReportMinutes, long nowMinutes)
    {
        var previous = previousReportMinutes.ToList();
        var day = previous.Count(m => nowMinutes - m < 24 * 60);
        var halfHour = previous.Any(m => nowMinutes - m < 30);
        return day < 3 && !halfHour;
    }

    public static readonly IReadOnlyList<string> DisputeReasons = ["didnt_happen", "wrong_place", "already_over", "duplicate"];
}
